File: app/src/main/java/projects/csce/evence/view/adapter/CardsAdapter.java
package projects.csce.evence.view.adapter;

import android.content.Context;
import android.content.res.Configuration;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;
import projects.csce.evence.R;
import projects.csce.evence.databinding.EventsListEntryLayoutBinding;
import daniel.perez.core.model.ViewCalendarData;
import daniel.perez.core.model.ViewEvent;

public class CardsAdapter extends RecyclerView.Adapter<CardsAdapter.ViewHolder> implements Observer<List<ViewCalendarData>>, Consumer<List<ViewCalendarData>>
{
    private static final String TAG = "CardsAdapter";
    private List<ViewCalendarData> dataList = Collections.emptyList();
    private Context context;
    private PublishSubject<ViewCalendarData> clicks = PublishSubject.create();

    public CardsAdapter(Context context)
    {
        this.context = context;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        EventsListEntryLayoutBinding binding = EventsListEntryLayoutBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull CardsAdapter.ViewHolder holder, int position) {
        holder.bindData(dataList.get(position));
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    @Override
    public void onChanged(List<ViewCalendarData> newData) {
        dataList = newData;
        notifyDataSetChanged();
    }

    public Observable<ViewCalendarData> clicks()
    {
        return clicks.debounce(300, TimeUnit.MILLISECONDS);
    }

    public void setData(List<ViewCalendarData> newData)
    {
        dataList = newData;
        notifyDataSetChanged();
    }

    @Override
    public void accept(List<ViewCalendarData> viewCalendarData)
    {
        setData(viewCalendarData);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        EventsListEntryLayoutBinding binding;

        ViewHolder(@NonNull EventsListEntryLayoutBinding binding)
        {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bindData(ViewCalendarData ical) {
            ical.getEvents().forEach(this::bind);

            binding.listEntryCardview.setOnClickListener(view -> clicks.onNext(ical) );
        }

        private void bind(ViewEvent event)
        {
            binding.listEntryTitleTextview.setText(event.getTitle());
            binding.listEntryDateTextview.setText(event.getStartDate());
            binding.listEntryTimeTextview.setText(event.getStartTime());
            binding.qrImageView.setImageBitmap(event.getImage());

            int isDark = context.getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;
            if (isDark == Configuration.UI_MODE_NIGHT_YES)
                binding.qrImageView.setColorFilter(ContextCompat.getColor(context, R.color.qr_dark_tint), android.graphics.PorterDuff.Mode.MULTIPLY);
            else if (isDark == Configuration.UI_MODE_NIGHT_NO)
                binding.qrImageView.clearColorFilter();

        }
    }
}

File: Core/src/main/java/teamevence/evenceapp/core/adapter/CardsAdapter.kt
package teamevence.evenceapp.core.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import teamevence.evenceapp.core.databinding.EventsListEntryLayoutBinding
import teamevence.evenceapp.core.db.timeString
import teamevence.evenceapp.core.model.UiPreference
import teamevence.evenceapp.core.model.ViewEvent
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.functions.Consumer
import io.reactivex.rxjava3.subjects.PublishSubject
import timber.log.Timber
import java.util.concurrent.TimeUnit

class CardsAdapter : RecyclerView.Adapter<CardsAdapter.ViewHolder>(), Consumer<List<ViewEvent>>
{
    private var dataList: MutableList<ViewEvent> = mutableListOf()
    private val clicks = PublishSubject.create<ViewEvent>()
    private var uiPreference: UiPreference? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = EventsListEntryLayoutBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindData(dataList[position])
    }

    override fun getItemCount(): Int {
        return dataList.size
    }

//    override fun onChanged(newData: List<ViewEvent>) {
//        val result = DiffUtil.calculateDiff( EventsDiffCallback(dataList, newData) )
//        result.dispatchUpdatesTo(this)
////        dataList = newData
////        notifyDataSetChanged()
//    }

    fun clicks(): Observable<ViewEvent> {
        return clicks.debounce(300, TimeUnit.MILLISECONDS)
    }

    fun updateUiFormat(uiPref: UiPreference?) {
        uiPreference = uiPref
    }

    fun setData(newData: List<ViewEvent>) {
        if(dataList.isEmpty())
        {
            dataList = newData.toMutableList()
            notifyDataSetChanged()
        }
        else
        {
            Timber.i("Diffing...")
            val result = DiffUtil.calculateDiff( EventsDiffCallback(dataList, newData) )
            dataList.clear()
            dataList.addAll(newData)
            result.dispatchUpdatesTo(this)
        }
    }

    override fun accept(viewCalendarData: List<ViewEvent>) {
        setData(viewCalendarData)
    }

    inner class ViewHolder(var binding: EventsListEntryLayoutBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bindData(event: ViewEvent) {
            bind(event)
            binding.listEntryCardview.setOnClickListener { view: View? -> clicks.onNext(event) }
        }

        private fun bind(event: ViewEvent) {
            binding.listEntryTitleTextview.text = event.title
            binding.listEntryPreviewDay.text = event.startDateTime.dayOfMonth.toString()
            binding.listEntryPreviewMonth.text = event.startDatePretty().substringBefore(" ")
            binding.listEntryTimeTextview.text = event.startDateTime.timeString()

        }
    }

    companion object {
        private const val TAG = "CardsAdapter"
    }
}